File: C/graphics.c
#include "graphics.h"
#include "stdio.h"
#include "string.h"
#include "SDL.h"
#include <sys/stat.h>


Vector* add_vec(Vector* v1, Vector* v2, Vector* res) {
    res->x = v1->x + v2->x;
    res->y = v1->y + v2->y;
    return res;
}

Vector* sub_vec(Vector* v1, Vector* v2, Vector* res) {
    res->x = v1->x - v2->x;
    res->y = v1->y - v2->y;
    return res;
}

long read_num(FILE* file, int offset, char size) {
    long res = 0;
    fseek(file, offset, SEEK_SET);
    fread(&res, size, 1, file);
    return res;
}

SDL_Surface* load_bmp(SDL_Renderer* renderer, const char* path) {
    FILE* asset = fopen(path,"rb");
    if (asset == NULL) return NULL;
    
    struct stat sb;
    if (stat(path, &sb) == -1) return NULL;
    if (read_num(asset, 0, 2) != 19778) return NULL;
    
    int w = read_num(asset, 18, 4);
    int h = read_num(asset, 22, 4);
    int pix = read_num(asset, 10, 4);
    int depth = read_num(asset, 28, 2);
    int file_size = read_num(asset, 2, 4);
    if (sb.st_size != file_size) return NULL;

    SDL_Surface* surf = SDL_CreateRGBSurfaceWithFormat(0,w,h,depth,SDL_PIXELFORMAT_BGR24);
    if (surf == NULL) return NULL;
    fseek(asset, pix, SEEK_SET);
    fread(surf->pixels, depth, w*h, asset);
    
    fclose(asset);
    return surf;
}

SDL_Texture* asset_to_texture(SDL_Renderer* renderer, const char* name) {
    char path[50] = "../assets/";
    strncat(path, name, 40);
    SDL_Surface* surf = load_bmp(renderer, path);
    if (surf == NULL) {
        printf("ERROR: failed to load asset \"%s\"\r\n", name);
        exit(1);
    }
    SDL_Texture* tex = SDL_CreateTextureFromSurface(renderer, surf);
    SDL_FreeSurface(surf);
    return tex;
}

GameAssets* generate_assets(SDL_Renderer* renderer) {
    GameAssets* assets = (GameAssets*) malloc(sizeof(GameAssets));
    assets->covered_tile = asset_to_texture(renderer,"covered_tile.bmp");
    assets->flagged_tile = asset_to_texture(renderer,"flagged_tile.bmp");
    assets->mine_tile = asset_to_texture(renderer,"mine_tile.bmp");
    assets->n_tiles[0] = asset_to_texture(renderer,"empty_tile.bmp");
    assets->n_tiles[1] = asset_to_texture(renderer,"one_tile.bmp");
    assets->n_tiles[2] = asset_to_texture(renderer,"two_tile.bmp");
    assets->n_tiles[3] = asset_to_texture(renderer,"three_tile.bmp");
    assets->n_tiles[4] = asset_to_texture(renderer,"four_tile.bmp");
    assets->n_tiles[5] = asset_to_texture(renderer,"five_tile.bmp");
    assets->n_tiles[6] = asset_to_texture(renderer,"six_tile.bmp");
    assets->n_tiles[7] = asset_to_texture(renderer,"seven_tile.bmp");
    assets->n_tiles[8] = asset_to_texture(renderer,"eight_tile.bmp");
    assets->n_tiles[9] = asset_to_texture(renderer,"nine_tile.bmp");
    return assets;
}

void free_assets(GameAssets* assets) {
    SDL_DestroyTexture(assets->covered_tile);
    SDL_DestroyTexture(assets->flagged_tile);
    SDL_DestroyTexture(assets->mine_tile);
    for (int i = 0; i < 10; i++)
        SDL_DestroyTexture(assets->n_tiles[i]);
    free(assets);
}

File: python/graphics.py
import pygame, sys

tileSize = 24
bannerHeight = 50
field_pos = (0,bannerHeight)
WHITE = pygame.Color("white")
PINK = pygame.Color("fuchsia")
RED = pygame.Color("red")
YELLOW = pygame.Color("yellow")
GREEN = pygame.Color("darkgreen")
CYAN = pygame.Color("cyan4")
BLUE = pygame.Color("blue")
GREY = pygame.Color("grey60")
DARK_GREEN = pygame.Color("darkgreen")
DARK_BLUE = pygame.Color("darkblue")
DARK_RED = pygame.Color("darkred")
LIGHT_GREY = pygame.Color("grey85")
DARK_GREY = pygame.Color("grey40")
BLACK = pygame.Color("black")

num_colors = [BLUE, DARK_GREEN, RED, DARK_BLUE, DARK_RED, CYAN, BLACK, DARK_GREY, PINK]

def add_vec(v1,v2): return (v1[0]+v2[0],v1[1]+v2[1])
def sub_vec(v1,v2): return (v1[0]-v2[0],v1[1]-v2[1])
def mul_vec(v,x): return (v[0]*x,v[1]*x)
def div_vec(v,x): return (v[0]//x,v[1]//x)

pygame.font.init()
font = pygame.font.SysFont('Comic Sans MS', 20)

empty_tile = pygame.Surface((tileSize,tileSize))
empty_tile.fill(LIGHT_GREY)
pygame.draw.rect(empty_tile,BLACK,empty_tile.get_rect(),1)

covered_tile = pygame.Surface((tileSize,tileSize))
covered_tile.fill(GREY)
pygame.draw.rect(covered_tile,DARK_GREY,covered_tile.get_rect(),3)
pygame.draw.rect(covered_tile,BLACK,covered_tile.get_rect(),1)

flagged_tile = pygame.Surface.copy(covered_tile)
pygame.draw.polygon(flagged_tile, RED, [(8,4),(17,8),(8,12)])
pygame.draw.line(flagged_tile, BLACK, (8,4), (8,19))

smiley_tile = pygame.Surface.copy(covered_tile)
pygame.draw.circle(smiley_tile, YELLOW, (12,12), 9)
pygame.draw.circle(smiley_tile, BLACK, (9,10), 1)
pygame.draw.circle(smiley_tile, BLACK, (15,10), 1)
pygame.draw.lines(smiley_tile, BLACK, 0, [(7,14),(10,16),(13,16),(16,14)])

frowny_tile = pygame.Surface.copy(covered_tile)
pygame.draw.circle(frowny_tile, YELLOW, (12,12), 9)
pygame.draw.line(frowny_tile, BLACK, (7,8),(9,10))
pygame.draw.line(frowny_tile, BLACK, (7,10),(9,8))
pygame.draw.line(frowny_tile, BLACK, (14,8),(16,10))
pygame.draw.line(frowny_tile, BLACK, (14,10),(16,8))
pygame.draw.circle(frowny_tile, BLACK, (12,16), 2)

mine_tile = pygame.Surface.copy(empty_tile)
pygame.draw.line(mine_tile, BLACK, (5,18), (18,5), 3)
pygame.draw.line(mine_tile, BLACK, (5,5), (18,18), 3)

n_tiles = [empty_tile]
for i in range(9):
    n_surf = font.render(str(i+1), True, num_colors[i])
    tile_surf = pygame.Surface.copy(empty_tile)
    tile_surf.blit(n_surf,div_vec(sub_vec((tileSize,tileSize-4),n_surf.get_size()),2))
    n_tiles.append(tile_surf)

if __name__ == "__main__":
    pygame.image.save(empty_tile, "assets/empty_tile.bmp")
    pygame.image.save(covered_tile, "assets/covered_tile.bmp")
    pygame.image.save(mine_tile, "assets/mine_tile.bmp")
    pygame.image.save(flagged_tile, "assets/flagged_tile.bmp")
    pygame.image.save(n_tiles[1], "assets/one_tile.bmp")
    pygame.image.save(n_tiles[2], "assets/two_tile.bmp")
    pygame.image.save(n_tiles[3], "assets/three_tile.bmp")
    pygame.image.save(n_tiles[4], "assets/four_tile.bmp")
    pygame.image.save(n_tiles[5], "assets/five_tile.bmp")
    pygame.image.save(n_tiles[6], "assets/six_tile.bmp")
    pygame.image.save(n_tiles[7], "assets/seven_tile.bmp")
    pygame.image.save(n_tiles[8], "assets/eight_tile.bmp")
    pygame.image.save(n_tiles[9], "assets/nine_tile.bmp")

    root = pygame.display.set_mode((500,100))
    root.fill(WHITE)

    x = 10
    root.blit(smiley_tile, (x,0)); x += 30
    root.blit(frowny_tile, (x,0)); x += 30
    root.blit(covered_tile, (x,0)); x += 30
    root.blit(flagged_tile, (x,0)); x += 30
    root.blit(mine_tile, (x,0)); x += 30
    for tile in n_tiles:
        root.blit(tile, (x,0)); x += 30

    while True:
        for event in pygame.event.get():
            if event.type == pygame.QUIT: sys.exit()
        pygame.display.update()